File: leetcode/409.longest-palindrome.py
# @leet start
class Solution:
    def longestPalindrome(self, s: str) -> int:
        counts: dict[str, int] = {}

        for ch in s:
            counts[ch] = counts.get(ch, 0) + 1

        odd = 0
        pairs = 0

        for count in counts.values():
            pairs += count // 2
            odd |= count % 2

        return pairs * 2 + odd


# @leet end

File: leetcode/409.longest-palindrome.ts
// @leet start
function longestPalindrome(s: string): number {
  const counts: Map<string, number> = new Map();
  let pairs = 0;
  let singles = 0;

  for (const c of s) {
    counts.set(c, 1 + (counts.get(c) ?? 0));

    if (counts.get(c)! % 2 === 0) {
      pairs++;
      singles--;
    } else {
      singles++;
    }
  }

  return 2 * pairs + Math.min(singles, 1);
}
// @leet end
